### ISSUE TITLE
Error compiling on kernel 4.10.2 - error: unknown type name 'cycle_t'

### ISSUE THREAD

**Issue by paolss** [ntop/PF_RING#193]
make[4]: Entering directory '/usr/src/linux-4.10.2'
  CC [M]  /usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.o
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5251:8: error: unknown type name 'cycle_t'
 static cycle_t e1000e_cyclecounter_read(const struct cyclecounter *cc)
        ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c: In function 'e1000e_cyclecounter_read':
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5257:2: error: unknown type name 'cycle_t'
  cycle_t systim, systim_next;
  ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5269:13: error: 'cycle_t' undeclared (first use in this function)
   systim = (cycle_t)systimel_1;
             ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5269:13: note: each undeclared identifier is reported only once for each function it appears in
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5269:21: error: expected ';' before 'systimel_1'
   systim = (cycle_t)systimel_1;
                     ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5270:22: error: expected ';' before 'systimeh'
   systim |= (cycle_t)systimeh << 32;
                      ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5274:21: error: expected ';' before 'systimel_2'
   systim = (cycle_t)systimel_2;
                     ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5275:22: error: expected ';' before 'systimeh'
   systim |= (cycle_t)systimeh << 32;
                      ^
In file included from /usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:48:0:
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/e1000.h:688:19: error: expected ';' before '__er32'
 #define er32(reg) __er32(hw, E1000_##reg)
                   ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5289:27: note: in expansion of macro 'er32'
    systim_next = (cycle_t)er32(SYSTIML);
                           ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/e1000.h:688:19: error: expected ';' before '__er32'
 #define er32(reg) __er32(hw, E1000_##reg)
                   ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5290:28: note: in expansion of macro 'er32'
    systim_next |= (cycle_t)er32(SYSTIMH) << 32;
                            ^
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c: In function 'e1000_sw_init':
/usr/src/PF_RING/drivers/intel/e1000e/e1000e-3.2.7.1-zc/src/netdev.c:5342:20: warning: assignment from incompatible pointer type
   adapter->cc.read = e1000e_cyclecounter_read;
                    ^

Operating system: Linux Gentoo

**Comment by cardigliano** [comment]
We will port the drivers ASAP. Please note that we support Ubuntu LTS / Debian / CentOS, this is lower priority.

**Comment by szarea** [comment]
Any ETA on porting the drivers for Kernel 4.10.X?

**Comment by astrosmash** [comment]
They removed this typedef in 4.10. You may replace with u64 as temp resort.
Writing because encountered similar issue with ptp in mainline igb 5.3.5.4

**Comment by szarea** [comment]
Thank you for your response.
yes, replacing " cycle_t" with u64 got us a bit further but the main issue still exists.
The current PF_RING code is not supporting Kernel 4.10.X. We still have many compiler errors, such as the following:

Any ETA on porting the drivers for Kernel  4.10X? This will help us the finalize or elevation? Many thanks.

/root/PF_RING/drivers/intel/igb/igb-5.3.3.5-zc/src/igb_main.c:2650:21: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
  .ndo_set_vf_vlan = igb_ndo_set_vf_vlan,                                                                                                                       
make -C /lib/modules/4.10.7-1.el6.elrepo.x86_64/build SUBDIRS=/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src modules
make[1]: Entering directory `/usr/src/kernels/4.10.7-1.el6.elrepo.x86_64'                                                 
  CC [M]  /root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.o                                               
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:8: error: type defaults to ‘int’ in declaration of ‘DEFINE_PCI_DEVICE_TABLE’ [-Werror=implicit-int]
 static DEFINE_PCI_DEVICE_TABLE(ixgbe_pci_tbl) = {                                                                                                                       
        ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:1: warning: parameter names (without types) in function declaration
 static DEFINE_PCI_DEVICE_TABLE(ixgbe_pci_tbl) = {                                                                                       
 ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:1: error: function ‘DEFINE_PCI_DEVICE_TABLE’ is initialized like a variable
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:2: warning: braces around scalar initializer
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:2: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
In file included from /root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:30:0:
./include/linux/pci.h:768:2: error: field name not in record or union initializer
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                             
  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:768:2: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                                  
  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
In file included from ./include/linux/pci.h:35:0,
                 from /root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:30:
./include/linux/pci_ids.h:2586:30: error: invalid initializer
 #define PCI_VENDOR_ID_INTEL  0x8086                         
                              ^
./include/linux/pci.h:768:12: note: in expansion of macro ‘PCI_VENDOR_ID_INTEL’
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                           
            ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci_ids.h:2586:30: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
 #define PCI_VENDOR_ID_INTEL  0x8086                                                        
                              ^
./include/linux/pci.h:768:12: note: in expansion of macro ‘PCI_VENDOR_ID_INTEL’
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                           
            ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
In file included from /root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:30:0:
./include/linux/pci.h:768:34: error: field name not in record or union initializer
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                              
                                  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:768:34: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                                   
                                  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:768:44: warning: excess elements in scalar initializer
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                        
                                            ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:768:44: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \                                   
                                            ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:769:2: error: field name not in record or union initializer
  .subvendor = PCI_ANY_ID, .subdevice = PCI_ANY_ID, 0, 0                         
  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
./include/linux/pci.h:769:2: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
  .subvendor = PCI_ANY_ID, .subdevice = PCI_ANY_ID, 0, 0                              
  ^
/root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},                                                                    
   ^
In file included from ./include/linux/pci.h:20:0,
                 from /root/PF_RING/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:30:
./include/linux/mod_devicetable.h:16:20: warning: excess elements in scalar initializer
 #define PCI_ANY_ID (~0)                                                               
                    ^
...

**Comment by astrosmash** [comment]
> igb_ndo_set_vf_vlan

VF api was changed as well, you may refer to this patch to add vlan protocol:
http://dpdk.org/dev/patchwork/patch/16651/

Yeah, there were a lot of API changes introduced recently. Hope you will succeed fixing the code yourself just in case of urgency. Waiting for @ntop's team to update the hooks ;)

**Comment by f3d0r4** [comment]
Any update on the progress of this issue?

**Comment by cardigliano** [comment]
All our systems have older kernels as we use LTS systems, I wasn't able to install this kernel so far for doing the porting sorry.

**Comment by archmangler** [comment]
I'm seeing this issue compiling the intel ixgbe drivers as well on this system:

Ubuntu 16.04.3 LTS
4.11.0-14-generic #20~16.04.1-Ubuntu SMP
PF_RING-6.6.0
libpcap-1.7.4 (libpcap patched for pf_ring)

`make -C /lib/modules/4.11.0-14-generic/build SUBDIRS=/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src modules
make[1]: Entering directory '/usr/src/linux-headers-4.11.0-14-generic'
  CC [M]  /opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.o
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:8: error: type defaults to ‘int’ in declaration of ‘DEFINE_PCI_DEVICE_TABLE’ [-Werror=implicit-int]
 static DEFINE_PCI_DEVICE_TABLE(ixgbe_pci_tbl) = {
        ^
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:1: warning: parameter names (without types) in function declaration
 static DEFINE_PCI_DEVICE_TABLE(ixgbe_pci_tbl) = {
 ^
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:122:1: error: function ‘DEFINE_PCI_DEVICE_TABLE’ is initialized like a variable
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:2: warning: braces around scalar initializer
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},
  ^
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:2: note: (near initialization for ‘DEFINE_PCI_DEVICE_TABLE’)
In file included from /opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:30:0:
./include/linux/pci.h:770:2: error: field name not in record or union initializer
  .vendor = PCI_VENDOR_ID_##vend, .device = (dev), \
  ^
/opt/pf_ring/PF_RING-6.6.0/drivers/intel/ixgbe/ixgbe-4.1.5-zc/src/ixgbe_main.c:123:3: note: in expansion of macro ‘PCI_VDEVICE’
  {PCI_VDEVICE(INTEL, IXGBE_DEV_ID_82598), 0},
   ^
`

**Comment by cardigliano** [comment]
@archmangler my Ubuntu 16.04.3 LTS has kernel 4.4.0-98, are you running ubuntu server? How did you get kernel 4.11.0?
Please also note that latest PF_RING 7 (stable) includes ixgbe v.5.0.4, did you check it?
